opm/autodiff/ISTLSolverEbos.hpp

```diff
@@ -381,11 +622,255 @@ namespace Opm
+        void makeOverlapRowsInvalid(Matrix& ebosJacIgnoreOverlap) const
+        {
+            //value to set on diagonal
+            typedef Dune::FieldMatrix<Scalar, numEq, numEq >        MatrixBlockType;
+            MatrixBlockType diag_block(0.0);
+            for (int eq = 0; eq < numEq; ++eq)
+                diag_block[eq][eq] = 1.0e100;
+
+            //loop over precalculated overlap rows and columns
+            for (auto row = overlapRowAndColumns_.begin(); row != overlapRowAndColumns_.end(); row++ )
+            {
+                int lcell = row->first;
+                //diagonal block set to large value diagonal
+                ebosJacIgnoreOverlap[lcell][lcell] = diag_block;
+
+                //loop over off diagonal blocks in overlap row
+                for (auto col = row->second.begin(); col != row->second.end(); ++col)
+                {
+                    int ncell = *col;
+                    //zero out block
+                    ebosJacIgnoreOverlap[lcell][ncell] = 0.0;
+                }
+            }
+        }
+       // weights to make approxiate pressure equations
+      Vector getStorageWeights(){
+	Vector weights(rhs_->size()); 
+	BlockVector rhs(0.0);
+	rhs[pressureIndex] = 1.0;
+	int index = 0;
+	ElementContext elemCtx(simulator_);
+	const auto& vanguard = simulator_.vanguard();
+	auto elemIt = vanguard.gridView().template begin</*codim=*/0>();
+	const auto& elemEndIt = vanguard.gridView().template end</*codim=*/0>();
+	for (; elemIt != elemEndIt; ++elemIt) {
+	  const Element& elem = *elemIt;
+	  elemCtx.updatePrimaryStencil(elem);
+	  elemCtx.updatePrimaryIntensiveQuantities(/*timeIdx=*/0);
+	  Dune::FieldVector<Evaluation, numEq> storage;
+	  unsigned threadId = ThreadManager::threadId();
+	  simulator_.model().localLinearizer(threadId).localResidual().computeStorage(storage,elemCtx,/*spaceIdx=*/0, /*timeIdx=*/0);
+	  Scalar extrusionFactor =
+	    elemCtx.intensiveQuantities(0, /*timeIdx=*/0).extrusionFactor();
+	  Scalar scvVolume =
+	    elemCtx.stencil(/*timeIdx=*/0).subControlVolume(0).volume() * extrusionFactor;
+	  Scalar storage_scale = scvVolume / elemCtx.simulator().timeStepSize();
+	  MatrixBlockType block;
+	  int offset = 0;
+	  double pressure_scale = 50e5;
+	  for(int ii=0; ii< numEq; ++ii){
+	    for(int jj=0; jj< numEq; ++jj){
+	      //const auto& vec = storage[ii].derivative(jj);
+	      block[ii][jj] = storage[ii].derivative(jj)/storage_scale;
+	      if(jj==0){
+		block[ii][jj] *=pressure_scale;
+	      }
+	    }
+	  }
+	  BlockVector bweights;
+	  MatrixBlockType block_transpose = block.transpose();
+	  block_transpose.solve(bweights, rhs);
+	  bweights /=1000; // given normal desnistyies this scales weights to about 1
+	  weights[index] = bweights;
+	  ++index;
+	}
+	return weights;
+      }
+
+      void scaleEquationsAndVariables(Vector& weights){
+	// loop over primary variables
+	const auto& sol = simulator_.model().solution(0);
```

Comment:
Maybe we should document somewhere what we are doing here in matrix vector notation.